File: Python/Algorithms/1588.py
from itertools import starmap


class Solution:
    def mults(self, l):
        # knock out special cases first.
        if l < 3:
            return [1] * l
        if l == 3:
            return [2] * l
        # build list of odd numbers (signifying groups) and the 
        # max amount of time they can occur in a list with length l.
        odds_pairs = [(o, (1 + (l - o))) for o in range(1, l + 1, 2)]
        
        # first element gets 1 from all groups.
        res = [len(odds_pairs)]  # for first element
        
        mid = ((l // 2) + 1) if l % 2 else l // 2
        for idx in range(2, mid + 1):
            sum_ = 0
            # the crux of the logic. not really easy to explain 
            # in comments so I'll try to explain in README.md
            for odd, occur in odds_pairs:
                if idx < odd and occur > idx:
                    sum_ += idx
                else:
                    if odd <= occur:
                        sum_ += odd
                    else:
                        sum_ += occur
            res.append(sum_)
            
        # for odd lists, elements are mirrored until 
        # mid - 2, append them (in reverse order).
        if l % 2:
            for i in range(len(res) - 2, - 1, -1):
                res.append(res[i])
        # for even lists, elements are mirrored until 
        # mid - 1, append them (in reverse order)
        else:
            for i in range(len(res) - 1, -1, - 1):
                res.append(res[i])
        # could yield all instead of returning explicit list but meh
        return res
    
    def sumOddLengthSubarrays(self, arr: List[int]) -> int:
        multipliers = self.mults(len(arr))
        multfunc = int.__mul__
        return sum(starmap(multfunc, zip(arr, multipliers)))

File: Rust/Algorithms/1588.rs
impl Solution {
    pub fn get_mults(size: i32) -> Vec<i32> {
        let mut arr: Vec<i32> = Vec::with_capacity(size as usize);
        let mut odds_occurs: Vec<(i32, i32)> = Vec::new();
    
        // create Vector of odd numbers and the number they 
        // occur in a Vector of size size.
        let mut i = 1;
        while i <= size{
            odds_occurs.push((i, (1 + (size - i))));
            i += 2;
        }
        // first element is # of odds.
        arr.push(odds_occurs.len() as i32);
        
        // find mid
        let mid = match size % 2 == 1 {
            false => size / 2,
            true => (size / 2) + 1,
        };
        
        // Build (half) vector of multiples.
        for idx in 2..mid+1 {
            let mut sum = 0;
            
            for (odd, occur) in &odds_occurs{
                if (idx < *odd) && (*occur > idx){
                    sum += idx;
                }
                else {
                    if *odd <= *occur{
                        sum += *odd;
                    } else {
                        sum += *occur;
                    }
                }
            }
            arr.push(sum);
        }
        
        // Build the other half.
        if size % 2 == 1 {
            for i in (0..=(arr.len() - 2)).rev(){
                arr.push(arr[i]);
            }
        } else {
            for i in (0..=(arr.len() - 1)).rev(){
                arr.push(arr[i]);
            }            
        }
        // return
        arr
    }
    
    pub fn sum_odd_length_subarrays(arr: Vec<i32>) -> i32 {
        // Handle fast edge cases.
        match arr.as_slice() {
            [x] => return *x,
            [x, y] => return *x + *y,
            [x, y, z] => return 2 * (*x + *y + *z),
            _ => (),
        }
        
        // Get Vec of multiples and make sum.
        let mut sum = 0;
        let mults = Solution::get_mults(arr.len() as i32);
        for i in 0..arr.len(){
            sum += arr[i] * mults[i];
        }
        
        sum
    }
}
